python/ngen_cal/src/ngen/cal/ngen_hooks/ngen_output.py

```diff
@@ -1,95 +1,298 @@
         """
-            After each iteration, copy the old outputs for possible future
-            evaluation and inspection.
+        After each iteration, copy the old outputs for possible future
+        evaluation and inspection.
         """
         path = info.workdir
-        out_dir = path/f"output_{iteration}"
+        out_dir = path / f"output_{iteration}"
         Path.mkdir(out_dir)
         globs = []
-        globs.append( path.glob(self.runoff_pattern) )
-        globs.append( path.glob(self.lateral_pattern) )
-        globs.append( path.glob(self.terminal_pattern) )
-        globs.append( path.glob(self.coastal_pattern) )
+        globs.append(path.glob(self.runoff_pattern))
+        globs.append(path.glob(self.lateral_pattern))
+        globs.append(path.glob(self.terminal_pattern))
+        globs.append(path.glob(self.coastal_pattern))
         for g in globs:
             for f in g:
-                f.rename(out_dir/f.name)
-        rpath = path/Path(self.routing_output)
-        rpath.rename(out_dir/rpath.name)
+                f.rename(out_dir / f.name)
+        rpath = path / Path(self.routing_output)
+        rpath.rename(out_dir / rpath.name)
+
+
+def _read_csv_output_v1_no_time(filepath: Path) -> pd.DataFrame:
+    # header: ","(0, 'q')","(0, 'v')","(0, 'd')",..."
+    # row   : "2420800,0.0,0.0,0.0,..."
+    # n_columns = 1 + number of timesteps (`nts`) * 3
+    df = pd.read_csv(filepath, index_col=0)
+    df.index = df.index.map(lambda x: "wb-" + str(x))
+    df.index.name = "waterbody_code"
+    tuples = [_parse_column_csv_output_v1(x) for x in df.columns]
+    df.columns = pd.MultiIndex.from_tuples(
+        tuples, names=("simulation_hour", "variable_name")
+    )
+    return df
+
+
+def _parse_column_csv_output_v1(s: str) -> tuple[int, str]:
+    # example column: (0, 'q')
+    assert s.startswith("(") and s.endswith("')")
+    int_end = 0
+    char_start = 0
+    for i, c in enumerate(s[:-2]):
+        if c == ",":
+            assert int_end == 0
+            int_end = i
+        elif c == "'":
+            char_start = i
+            break
+
+    assert char_start > int_end
+    return int(s[1:int_end]), s[char_start + 1 : -2]
+
+
+def _routing_timestep_size_s(routing_n_ts: int, realization: NgenRealization) -> int:
+    """Routing timestep size in seconds."""
+    start = realization.time.start_time
+    end = realization.time.end_time
+    ngen_ts_s = realization.time.output_interval
+    # start is not included in interval b.c. first _ngen_ model output is start + dt
+    ngen_n_ts = len(pd.date_range(start, end, freq=f"{ngen_ts_s}s", inclusive="right"))
+
+    r_n_ts_in_ngen_ts = routing_n_ts / ngen_n_ts
+
+    r_dt, r = divmod(ngen_ts_s, r_n_ts_in_ngen_ts)
+    assert r == 0, "routing timestep is not evenly divisible by ngen_timesteps"
+    return int(r_dt)
+
+
+def _csv_output_v1(p: Path, realization: NgenRealization) -> _NgenCalModelOutputFn:
```

Comment:
One of the biggest concerns about using these as module level functions is the lack of conceptual encapsulation.  `ngen_output` is module for ngen output, not just t-route output, and may (probably should) extend its capabilities to use/provide other ngen model outputs, which could get a bit messy having these as module scoped funtions with seemingly generic names.

Reply:
Yeah I can empathize with the names feeling a little generic. I mirrored the naming with how they are named in `t-route`'s config (see [here](https://github.com/NOAA-OWP/ngen-cal/issues/153#issuecomment-2274204258) for more info). So for example, this function maps to a `t-route` config that looks like:

```yaml
output_parameters:
  csv_output:
    csv_output_folder: output/
```

Other example:

```yaml
output_parameters:
  stream_output: 
      stream_output_directory: output/
      stream_output_time: 10000000 # [hr]
      stream_output_type: '.csv'
      stream_output_internal_frequency: 60
```